// provisioner/lib/recipes/apps/etcd/add_environment_variables.js
/// <reference path="../../../../typings-recipe.d.ts" />
/*
 * Add environment variables to shell sessions through an env file
 */
"use strict";
recipes.register({
    id: "etcd-environment",
    on: { beforeInitialize: {} },
    conditions: {
        tierModules: { any: ["etcd"] },
    },
    recipeHandler: function (input) {
        const envFile = path.join(platform.pathInfo.namiAppPath, "etcd", ".deployment.env");
        const envVars = {
            ETCDCTL_API: 3,
        };
        fs.writeFileSync(envFile, "");
        for (const envVar in envVars) {
            fs.appendFileSync(envFile, `export ${envVar}="${envVars[envVar]}"\n`);
        }
        for (let path of glob.sync("{/root/{.bashrc,.profile},/home/*/{.bashrc,.profile},/etc/skel/{.bashrc,.profile}}")) {
            // This loop relies on the creation of the files by 'add-to-system-path' recipe executed in a previous hook
            platform.createFileSection(path, {
                title: "ETCD ENVIRONMENT SETTINGS",
                contents: `\nsource ${envFile}\n`
            });
        }
    }
});
