// provisioner/lib/recipes/apps/airflow/add_environment_variables.js
/// <reference path="../../../../typings-recipe.d.ts" />
/*
 * Add environment variables to shell sessions through an env file
 */
"use strict";
recipes.register({
    id: "airflow-environment",
    on: { beforeInitialize: {} },
    conditions: {
        tierModules: { any: ["airflow", "airflow-scheduler", "airflow-worker"] },
    },
    recipeHandler: function (input) {
        const installdir = path.join(platform.pathInfo.namiAppPath, "airflow");
        const envFile = path.join(installdir, ".deployment.env");
        const pythonInstalldir = path.join(platform.pathInfo.namiAppPath, "python");
        const envVars = {
            AIRFLOW_HOME: installdir,
            LD_LIBRARY_PATH: `${path.join(pythonInstalldir, "lib")}:`
                + `${path.join(installdir, "venv/lib/python3.8/site-packages/numpy.libs")}:$LD_LIBRARY_PATH`,
        };
        fs.writeFileSync(envFile, "");
        for (const envVar in envVars) {
            fs.appendFileSync(envFile, `export ${envVar}="${envVars[envVar]}"\n`);
        }
        for (let path of glob.sync("{/root/{.bashrc,.profile},/home/*/{.bashrc,.profile},/etc/skel/{.bashrc,.profile}}")) {
            // This loop relies on the creation of the files by 'add-to-system-path' recipe executed in a previous hook
            platform.createFileSection(path, {
                title: "AIRFLOW ENVIRONMENT SETTINGS",
                contents: `\nsource ${envFile}\n`
            });
        }
    }
});
