// provisioner/lib/recipes/apps/java/add_environment_variables.js
/// <reference path="../../../../typings-recipe.d.ts" />
/*
 * Add environment variables to shell sessions through an env file
 */
"use strict";
recipes.register({
    id: "java-environment",
    on: { beforeInitialize: {} },
    conditions: {
        tierModules: { any: ["java"] },
    },
    recipeHandler: function (input) {
        const installdir = path.join(platform.pathInfo.namiAppPath, "java");
        const envFile = path.join(installdir, ".deployment.env");
        const envVars = {
            JAVA_HOME: installdir,
        };
        fs.writeFileSync(envFile, "");
        for (const envVar in envVars) {
            fs.appendFileSync(envFile, `export ${envVar}="${envVars[envVar]}"\n`);
        }
        for (let path of glob.sync("{/root/{.bashrc,.profile},/home/*/{.bashrc,.profile},/etc/skel/{.bashrc,.profile}}")) {
            // This loop relies on the creation of the files by 'add-to-system-path' recipe executed in a previous hook
            platform.createFileSection(path, {
                title: "JAVA ENVIRONMENT SETTINGS",
                contents: `\nsource ${envFile}\n`
            });
        }
    }
});
